<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>An alternative method for the building of a RepSeqExperiment object — readFormatSet • AnalyzAIRR</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="An alternative method for the building of a RepSeqExperiment object — readFormatSet"><meta name="description" content="This function allows the loading of alignment files in other formats than the ones supported by the readAIRRSet function.
These files should contain the following column names:
- sample_id: Sample names
- ntCDR3: nucleotide CDR3 sequence
- aaCDR3: amino acid CDR3 sequence
- V: Variable gene name (IMGT nomenclature)
- J: Joining gene name (IMGT nomenclature)
- count: the occurrence of each sequence
The order of the columns must be respected. Input files can contain additional columns that won't however be taken into account in the generated RepSeqExperiment object.
Similarly to the readAIRRSet function, readFormatSet applies a series of filters during the building process.
The possible filters that can be applied include:
- filtering out ambiguous sequences
- filtering out unproductive sequences including the ones with stop codons, namely out-of-frame sequences
- filtering out singletons, i.e clonotypes with an occurrence of 1
- filtering out short or extensively long amino acid CDR3 sequences"><meta property="og:description" content="This function allows the loading of alignment files in other formats than the ones supported by the readAIRRSet function.
These files should contain the following column names:
- sample_id: Sample names
- ntCDR3: nucleotide CDR3 sequence
- aaCDR3: amino acid CDR3 sequence
- V: Variable gene name (IMGT nomenclature)
- J: Joining gene name (IMGT nomenclature)
- count: the occurrence of each sequence
The order of the columns must be respected. Input files can contain additional columns that won't however be taken into account in the generated RepSeqExperiment object.
Similarly to the readAIRRSet function, readFormatSet applies a series of filters during the building process.
The possible filters that can be applied include:
- filtering out ambiguous sequences
- filtering out unproductive sequences including the ones with stop codons, namely out-of-frame sequences
- filtering out singletons, i.e clonotypes with an occurrence of 1
- filtering out short or extensively long amino acid CDR3 sequences"><meta property="og:image" content="https://i3-unit.github.io/AnalyzAIRR/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">AnalyzAIRR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorial" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorial</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorial"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Getting started</h6></li>
    <li><a class="dropdown-item" href="../articles/gettingstarted.html">Installation</a></li>
    <li><a class="dropdown-item" href="../articles/dataloading.html">Data loading</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Comprehensive data exploration</h6></li>
    <li><a class="dropdown-item" href="../articles/exploratory.html">Exploratory analysis</a></li>
    <li><a class="dropdown-item" href="../articles/onesample.html">One-sample analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Advanced data manipulation</h6></li>
    <li><a class="dropdown-item" href="../articles/datamanipulation.html">Data filtering</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>In-depth cross-sample comparison</h6></li>
    <li><a class="dropdown-item" href="../articles/datanormalization.html">Data normalization</a></li>
    <li><a class="dropdown-item" href="../articles/multisample.html">Multi-sample analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Interactive AIRR data analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/shiny.html">Shiny-AnalyzAIRR</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/i3-unit/AnalyzAIRR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>An alternative method for the building of a RepSeqExperiment object</h1>
      <small class="dont-index">Source: <a href="https://github.com/i3-unit/AnalyzAIRR/blob/master/R/loadData.R" class="external-link"><code>R/loadData.R</code></a></small>
      <div class="d-none name"><code>readFormatSet.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function allows the loading of alignment files in other formats than the ones supported by the <code><a href="readAIRRSet.html">readAIRRSet</a></code> function.
These files should contain the following column names:
- sample_id: Sample names</p>
<p>- ntCDR3: nucleotide CDR3 sequence</p>
<p>- aaCDR3: amino acid CDR3 sequence</p>
<p>- V: Variable gene name (IMGT nomenclature)</p>
<p>- J: Joining gene name (IMGT nomenclature)</p>
<p>- count: the occurrence of each sequence</p>
<p>The order of the columns must be respected. Input files can contain additional columns that won't however be taken into account in the generated RepSeqExperiment object.</p>
<p>Similarly to the <code><a href="readAIRRSet.html">readAIRRSet</a></code> function, <code>readFormatSet</code> applies a series of filters during the building process.
The possible filters that can be applied include:</p>
<p>- filtering out ambiguous sequences</p>
<p>- filtering out unproductive sequences including the ones with stop codons, namely out-of-frame sequences</p>
<p>- filtering out singletons, i.e clonotypes with an occurrence of 1</p>
<p>- filtering out short or extensively long amino acid CDR3 sequences</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">readFormatSet</span><span class="op">(</span></span>
<span>  <span class="va">fileList</span>,</span>
<span>  chain <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRA"</span>, <span class="st">"TRB"</span>, <span class="st">"TRG"</span>, <span class="st">"TRD"</span>, <span class="st">"IGH"</span>, <span class="st">"IGK"</span>, <span class="st">"IGL"</span><span class="op">)</span>,</span>
<span>  sampleinfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  keep.ambiguous <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  keep.unproductive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  filter.singletons <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  aa.th <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  outFiltered <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">1L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-filelist">fileList<a class="anchor" aria-label="anchor" href="#arg-filelist"></a></dt>
<dd><p>a list of paths to the alignment files. File format can be one of the following: .tsv, txt.gz, .zip, or .tar</p></dd>


<dt id="arg-chain">chain<a class="anchor" aria-label="anchor" href="#arg-chain"></a></dt>
<dd><p>a character vector indicating a single TCR or Ig chain to analyze. The vector can be one of the following:</p>
<p>- "TRA", "TRB","TRG" or "TRD" for the TCR repertoires</p>
<p>- "IGH","IGK" or "IGL" for the BCR repertoires</p></dd>


<dt id="arg-sampleinfo">sampleinfo<a class="anchor" aria-label="anchor" href="#arg-sampleinfo"></a></dt>
<dd><p>a data frame containing:</p>
<p>- a column with the sample_ids. Ids should match the base names of the corresponding files as well as their order in the list. This column should be assigned as row.names when the metadata file is loaded (see the example below).</p>
<p>- any additional columns with relevant information for the analyses. Group columns must be transformed into factors after loading (see the example below).</p>
<p>No specific column names nor order are required</p></dd>


<dt id="arg-keep-ambiguous">keep.ambiguous<a class="anchor" aria-label="anchor" href="#arg-keep-ambiguous"></a></dt>
<dd><p>a boolean indicating whether or not ambiguous sequences containing an "N" in their nucleotide CDR3 should be left in. Default is <code>FALSE</code>.</p></dd>


<dt id="arg-keep-unproductive">keep.unproductive<a class="anchor" aria-label="anchor" href="#arg-keep-unproductive"></a></dt>
<dd><p>a boolean indicating whether or not unproductive sequences should be left in. Default is <code>FALSE</code>.
Unproductive sequences include:</p>
<p>- out-of-frame sequences: sequences with frame shifts based on the number of nucleotides in their CDR3s</p>
<p>- sequences containing stop codons: aaCDR3 sequences with a "*" or "~" should be kept. Default is <code>FALSE</code></p></dd>


<dt id="arg-filter-singletons">filter.singletons<a class="anchor" aria-label="anchor" href="#arg-filter-singletons"></a></dt>
<dd><p>a boolean indicating whether or not ntClones (V+ntCDR3+J) with an occurrence of 1 should be filtered out. Default is <code>FALSE</code></p></dd>


<dt id="arg-aa-th">aa.th<a class="anchor" aria-label="anchor" href="#arg-aa-th"></a></dt>
<dd><p>an integer determining the CDR3 amino acid sequence length limits, i.e. the maximum number of amino acids deviating from the mean length that is accepted. The default value is 8, which keeps amino acid CDR3s with a length that falls inside the following range: mean length-8 =&lt; aa.th &lt;= mean length+ 8.</p></dd>


<dt id="arg-outfiltered">outFiltered<a class="anchor" aria-label="anchor" href="#arg-outfiltered"></a></dt>
<dd><p>a boolean indicating whether or not to write in the oData slot of the RepSeqexperiment object a data frame containing the filtered out reads.</p></dd>


<dt id="arg-cores">cores<a class="anchor" aria-label="anchor" href="#arg-cores"></a></dt>
<dd><p>an integer indicating the number of cores to be uses by the function. The process is paralleled and can thus be used on multiple cores. Default is 1.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>an object of class <code><a href="RepSeqExperiment-class.html">RepSeqExperiment</a></code> that is  used in all the analytical metrics proposed by the AnalyzAIRR package. See <code><a href="RepSeqExperiment-class.html">RepSeqExperiment-class</a></code> for more details.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">'extdata/informat'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                     package <span class="op">=</span> <span class="st">'AnalyzAIRR'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                     full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vanessa Mhanna, Gabriel Pires, Grégoire Bohl, Karim El Soufi, Nicolas Tchitchek, David Klatzmann, Adrien Six, Hang-Phuong Pham, Encarnita Mariotti-Ferrandiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

