<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Data loading • AnalyzAIRR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data loading">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">AnalyzAIRR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorial" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorial</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorial">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Getting started</h6></li>
    <li><a class="dropdown-item" href="../articles/gettingstarted.html">Installation</a></li>
    <li><a class="dropdown-item" href="../articles/dataloading.html">Data loading</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Comprehensive data exploration</h6></li>
    <li><a class="dropdown-item" href="../articles/exploratory.html">Exploratory analysis</a></li>
    <li><a class="dropdown-item" href="../articles/onesample.html">One-sample analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Advanced data manipulation</h6></li>
    <li><a class="dropdown-item" href="../articles/datamanipulation.html">Data filtering</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>In-depth cross-sample comparison</h6></li>
    <li><a class="dropdown-item" href="../articles/datanormalization.html">Data normalization</a></li>
    <li><a class="dropdown-item" href="../articles/multisample.html">Multi-sample analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Interactive AIRR data analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/shiny.html">Shiny-AnalyzAIRR</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/i3-unit/AnalyzAIRR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="dataloading_files/core-js-2.5.3/shim.min.js"></script><script src="dataloading_files/react-18.2.0/react.min.js"></script><script src="dataloading_files/react-18.2.0/react-dom.min.js"></script><script src="dataloading_files/reactwidget-2.0.0/react-tools.js"></script><link href="dataloading_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="dataloading_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="dataloading_files/reactable-0.4.4/reactable.css" rel="stylesheet">
<script src="dataloading_files/reactable-binding-0.4.4/reactable.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Data loading</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/i3-unit/AnalyzAIRR/blob/HEAD/vignettes/dataloading.Rmd" class="external-link"><code>vignettes/dataloading.Rmd</code></a></small>
      <div class="d-none name"><code>dataloading.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="input-format">Input format<a class="anchor" aria-label="anchor" href="#input-format"></a>
</h2>
<p><strong>AnalyzAIRR</strong> proposes two different data loading
processes depending on the alignment files format.</p>
<div class="section level3">
<h3 id="automatic-format-detection">Automatic format detection<a class="anchor" aria-label="anchor" href="#automatic-format-detection"></a>
</h3>
<p><strong>AnalyzAIRR</strong> integrates parser functions to
automatically detect the following formats:</p>
<ul>
<li><p><strong>MiXCR</strong>: <a href="https://github.com/milaboratory/mixcr" class="external-link uri">https://github.com/milaboratory/mixcr</a></p></li>
<li><p><strong>immunoSEQ</strong>: <a href="https://www.adaptivebiotech.com/adaptive-immunosequencing/" class="external-link uri">https://www.adaptivebiotech.com/adaptive-immunosequencing/</a></p></li>
<li><p><strong>MiAIRR</strong>: <a href="https://docs.airr-community.org/en/latest/datarep/rearrangements.html" class="external-link uri">https://docs.airr-community.org/en/latest/datarep/rearrangements.html</a></p></li>
</ul>
<p>These alignment files can be loaded using the
<code><a href="../reference/readAIRRSet.html">readAIRRSet()</a></code> function.</p>
</div>
<div class="section level3">
<h3 id="reformatting-of-annotated-files">Reformatting of annotated files<a class="anchor" aria-label="anchor" href="#reformatting-of-annotated-files"></a>
</h3>
<p><strong>AnalyzAIRR</strong> offers the possibility to load alignment
files in other formats than the ones cited in the previous section.
These files should contain the following required column names:</p>
<ul>
<li><p><strong>sample_id</strong>: Sample names</p></li>
<li><p><strong>V</strong>: Variable gene name (IMGT
nomenclature)</p></li>
<li><p><strong>J</strong>: Joining gene name (IMGT
nomenclature)</p></li>
<li><p><strong>ntCDR3</strong>: Nucleotide CDR3 sequence</p></li>
<li><p><strong>aaCDR3</strong>: Amino acid CDR3 sequence</p></li>
<li><p><strong>count</strong>: The occurrence of each sequence within
the sample</p></li>
</ul>
<p><strong>Note</strong>: The order of the columns must be respected.
Input files can contain additional columns that will not be taken into
account during the loading process.</p>
<p><code><a href="../reference/readFormatSet.html">readFormatSet()</a></code> is the equivalent function of
<code><a href="../reference/readAIRRSet.html">readAIRRSet()</a></code> when using user-formatted files. It allows
the loading and filtering of a list of files using the same previously
described parameters.</p>
</div>
</div>
<div class="section level2">
<h2 id="data-loading">Data loading<a class="anchor" aria-label="anchor" href="#data-loading"></a>
</h2>
<div class="section level3">
<h3 id="paths-to-input-files">Paths to input files<a class="anchor" aria-label="anchor" href="#paths-to-input-files"></a>
</h3>
<p>Specify the list of paths to the aligned files.</p>
<p>A list of 8 samples were selected from a published study (Mhanna et
al., 2021) to illustrate the use of this package. Samples were collected
from the spleen of 4 healthy mice and sorted into naïve regulatory T
cells (nTreg), and activated/memory regulatory T cells (amTreg). Fastq
files were aligned using MiXCR and exported in a MiAIRR format, and can
be found at <a href="https://github.com/i3-unit/AnalyzAIRR/tree/master/inst/extdata/MiAIRR" class="external-link uri">https://github.com/i3-unit/AnalyzAIRR/tree/master/inst/extdata/MiAIRR</a></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">'extdata/MiAIRR'</span><span class="op">)</span>,</span>
<span>                            package <span class="op">=</span> <span class="st">'AnalyzAIRR'</span><span class="op">)</span>,</span>
<span>                            full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="metadata">Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p>It is possible to provide a metadata if users wish to perform
inter-group comparative analyses.</p>
<p>The metadata should be provided as a dataframe containing:</p>
<ul>
<li>a column with the sample names that match the name of the alignment
files and their order in the list.</li>
</ul>
<p>Only one column containing the sample names should be provided in the
metadata file. This column is assigned to the <code>row.names</code>
argument when loading the metadata.</p>
<ul>
<li>any additional columns with relevant information for the analyses.
Columns could encompass the experimental conditions, clinical variables,
etc…</li>
</ul>
<p>No specific column names nor order are required.</p>
<p><strong>Note</strong>: Since <strong>R 4.0.0</strong>,
<code>stringsAsFactors</code> is set to <strong>FALSE</strong> by
default, user will thus need to coerce character vectors into
factors.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">'extdata/sampledata.txt'</span><span class="op">)</span>,</span>
<span>                            package <span class="op">=</span> <span class="st">'AnalyzAIRR'</span><span class="op">)</span>, </span>
<span>                            header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">cell_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">cell_subset</span><span class="op">)</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">sex</span><span class="op">)</span></span></code></pre></div>
<div class="reactable html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"tag":{"name":"Reactable","attribs":{"data":{".rownames":["tripod-30-813","tripod-30-815","tripod-31-846","tripod-31-848","tripod-35-970","tripod-35-972","tripod-36-1003","tripod-36-1005"],"cell_subset":["amTreg","nTreg","amTreg","nTreg","amTreg","nTreg","nTreg","amTreg"],"sex":["F","F","M","M","F","F","M","M"]},"columns":[{"id":".rownames","name":"","type":"character","align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"},"sortable":false,"filterable":false,"rowHeader":true},{"id":"cell_subset","name":"cell_subset","type":"factor","align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"sex","name":"sex","type":"factor","align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}}],"bordered":true,"dataKey":"64afe0fa2b187f7ee43e49f1b62343cb"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="loading">Loading<a class="anchor" aria-label="anchor" href="#loading"></a>
</h3>
<p>Whether you’re using the <code><a href="../reference/readAIRRSet.html">readAIRRSet()</a></code> or
<code><a href="../reference/readFormatSet.html">readFormatSet()</a></code> function, the following parameters must be
specified:</p>
<ul>
<li><p><code>fileList</code>: The list of paths to the input files. File
extensions can be one of the following: .tsv, txt.gz, .zip,
.tar</p></li>
<li><p><code>fileFormat</code>: The name of the tool that was used to
generate the alignment files. Should be one of “MiXCR”, “immunoseq” or
“MiAIRR”.</p></li>
<li>
<p><code>chain</code>: The chain type to be analyzed. Should be one
of the following:</p>
<ul>
<li>“TRA”, “TRB”,“TRG” or “TRD” for the TCR repertoires</li>
<li>“IGH”,“IGK” or “IGL” for the BCR repertoires</li>
</ul>
</li>
</ul>
<p>Only a single chain can be loaded at once.</p>
<ul>
<li><p><code>sampleinfo</code>: A data frame with the metadata
information</p></li>
<li><p><code>keep.ambiguous</code>: A choice whether ambiguous sequences
should be left in the datasets</p></li>
<li><p><code>keep.unproductive</code>: A choice whether unproductive
sequences should be left in the datasets</p></li>
<li><p><code>filter.singletons</code>: A choice whether sequences with
an occurrence of 1 should be filtered out</p></li>
<li><p><code>aa.th</code>: A threshold determining the amino acid CDR3
sequence length limits. Refer to the function’s documentation for more
details</p></li>
<li><p><code>outFiltered</code>: A choice whether to write an output
file containing the filtered out reads by the previously cited filtering
parameters</p></li>
<li><p><code>cores</code>: The number of CPU cores to be used for a
parallel processing</p></li>
</ul>
<p>It is possible to apply the above-mentioned filtering functions
subsequently to the data loading step in case users wish to explore the
datasets before any data manipulation.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RepSeqData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readAIRRSet.html">readAIRRSet</a></span><span class="op">(</span>fileList <span class="op">=</span> <span class="va">l</span>,</span>
<span>                          fileFormat <span class="op">=</span> <span class="st">"MiAIRR"</span>,</span>
<span>                          chain <span class="op">=</span> <span class="st">"TRA"</span>,</span>
<span>                          sampleinfo <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>                          keep.ambiguous <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          keep.unproductive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          filter.singletons <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          aa.th <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                          outFiltered <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          cores <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="the-repseqexperiment-object">The RepSeqExperiment object<a class="anchor" aria-label="anchor" href="#the-repseqexperiment-object"></a>
</h2>
<div class="section level3">
<h3 id="architecture">Architecture<a class="anchor" aria-label="anchor" href="#architecture"></a>
</h3>
<p>Data loading with either one of <code><a href="../reference/readFormatSet.html">readFormatSet()</a></code> or
<code><a href="../reference/readAIRRSet.html">readAIRRSet()</a></code> allows the generation of a RepSeqExperiment
object. This object is used as input in all of the functions.</p>
<p>It is composed of 4 slots, each containing a different type of
information.</p>
<ul>
<li>
<p><strong>assayData</strong>: a data.table composed of all the
alignment datasets and containing the following columns:</p>
<ul>
<li><p><strong>sample_id</strong>: Sample names</p></li>
<li><p><strong>V</strong>: Variable gene name</p></li>
<li><p><strong>J</strong>: Joining gene name</p></li>
<li><p><strong>VJ</strong>: V-J gene combination using V and J gene
names.</p></li>
<li><p><strong>ntCDR3</strong>: Nucleotide CDR3 sequence</p></li>
<li><p><strong>aaCDR3</strong>: Amino acid CDR3 sequence</p></li>
<li><p><strong>aaClone</strong>: The full sequence including the V gene,
the amino acid CDR3 sequence and the J gene</p></li>
<li><p><strong>ntClone</strong>: The full sequence including the V gene,
the nucleotide CDR3 sequence and the J gene</p></li>
<li><p><strong>count</strong>: The occurrence of each clone</p></li>
</ul>
</li>
</ul>
<p>This slot can be extracted using the <code><a href="../reference/RepSeqExperiment-class.html">assay()</a></code>
function.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RepSeqExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">RepSeqData</span><span class="op">)</span></span></code></pre></div>
<div id="htmlwidget-e5c8c404fe174e4c81bd" class="reactable html-widget" style="width:auto;height:300px;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"sample_id":["tripod-30-813","tripod-30-813","tripod-30-813","tripod-30-813","tripod-30-813","tripod-30-813"],"V":["TRAV14N-3","TRAV8-1","TRAV5-4","TRAV5D-4","TRAV6-6","TRAV7-2"],"J":["TRAJ16","TRAJ18","TRAJ31","TRAJ13","TRAJ12","TRAJ23"],"VJ":["TRAV14N-3 TRAJ16","TRAV8-1 TRAJ18","TRAV5-4 TRAJ31","TRAV5D-4 TRAJ13","TRAV6-6 TRAJ12","TRAV7-2 TRAJ23"],"ntCDR3":["TGTGCAGCAAGTAGGGCTTCAAGTGGCCAGAAGCTGGTTTTT","TGTGCTACTGATGGAAGAGGTTCAGCCTTAGGGAGGCTGCATTTT","TGTGCTGCAAGTAATAGCAATAACAGAATCTTCTTT","TGTGCTGCAGCGGATTCTGGGACTTACCAGAGGTTT","TGTGCTCTGGGTGAGCCTGGAGGCTATAAAGTGGTCTTT","TGTGCAGCCAGGGGGAATTATAACCAGGGGAAGCTTATCTTT"],"aaCDR3":["CAASRASSGQKLVF","CATDGRGSALGRLHF","CAASNSNNRIFF","CAAADSGTYQRF","CALGEPGGYKVVF","CAARGNYNQGKLIF"],"aaClone":["TRAV14N-3 CAASRASSGQKLVF TRAJ16","TRAV8-1 CATDGRGSALGRLHF TRAJ18","TRAV5-4 CAASNSNNRIFF TRAJ31","TRAV5D-4 CAAADSGTYQRF TRAJ13","TRAV6-6 CALGEPGGYKVVF TRAJ12","TRAV7-2 CAARGNYNQGKLIF TRAJ23"],"ntClone":["TRAV14N-3 TGTGCAGCAAGTAGGGCTTCAAGTGGCCAGAAGCTGGTTTTT TRAJ16","TRAV8-1 TGTGCTACTGATGGAAGAGGTTCAGCCTTAGGGAGGCTGCATTTT TRAJ18","TRAV5-4 TGTGCTGCAAGTAATAGCAATAACAGAATCTTCTTT TRAJ31","TRAV5D-4 TGTGCTGCAGCGGATTCTGGGACTTACCAGAGGTTT TRAJ13","TRAV6-6 TGTGCTCTGGGTGAGCCTGGAGGCTATAAAGTGGTCTTT TRAJ12","TRAV7-2 TGTGCAGCCAGGGGGAATTATAACCAGGGGAAGCTTATCTTT TRAJ23"],"count":[1942,1577,1348,882,796,779]},"columns":[{"id":"sample_id","name":"sample_id","type":"character","minWidth":200,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"V","name":"V","type":"character","minWidth":100,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"J","name":"J","type":"character","minWidth":100,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"VJ","name":"VJ","type":"character","minWidth":100,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"ntCDR3","name":"ntCDR3","type":"character","minWidth":200,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"aaCDR3","name":"aaCDR3","type":"character","minWidth":200,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"aaClone","name":"aaClone","type":"character","minWidth":200,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"ntClone","name":"ntClone","type":"character","minWidth":200,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"count","name":"count","type":"numeric","minWidth":100,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}}],"pagination":false,"bordered":true,"height":"300px","dataKey":"af1b9aef2f9fd8cabd2843be703dda7a"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script><ul>
<li>
<p><strong>metaData</strong>: a data frame containing the metadata
information provided for the building of the RepSeqExperiment object,
followed by a number of statistics calculated for each sample.</p>
<p>These statistics include:</p>
<ul>
<li><p><strong>nSequences</strong>: the total number of sequences in a
sample</p></li>
<li><p><strong>V</strong>: The total number of V genes expressed in each
sample</p></li>
<li><p><strong>J</strong>: The total number of J genes</p></li>
<li><p><strong>VJ</strong>: The total number of V-J gene
combinations</p></li>
<li><p><strong>ntCDR3</strong>: The number of unique nucleotide
CDR3s</p></li>
<li><p><strong>aaCDR3</strong>: The number of unique amino acid
CDR3s</p></li>
<li><p><strong>aaClone</strong>: The number of unique aaClones</p></li>
<li><p><strong>ntClone</strong>: The number of unique ntClones</p></li>
<li><p><strong>Chao1</strong>: Estimates total species richness using
the counts of rare species (singletons and doubletons) (Chao, 1984). A
higher Chao1 value indicates greater estimated species richness. It
corrects for unseen species by extrapolating from rare species
counts.</p></li>
<li><p><strong>Improved Chao1</strong>: An extension of
<strong>Chao1</strong> which uses higher-order rare species,
specifically, tripletons and quadrupletons, and adjusts for sample size
to improve accuracy in small or unevenly samples datasets (Chiu et al.,
2014).</p></li>
</ul>
</li>
</ul>
<p>This slot can be extracted using the <code><a href="../reference/RepSeqExperiment-class.html">mData()</a></code>
function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RepSeqExperiment-class.html">mData</a></span><span class="op">(</span><span class="va">RepSeqData</span><span class="op">)</span></span></code></pre></div>
<div class="reactable html-widget html-fill-item" id="htmlwidget-36aa3d2a04d42bbc2145" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-36aa3d2a04d42bbc2145">{"x":{"tag":{"name":"Reactable","attribs":{"data":{".rownames":["tripod-30-813","tripod-30-815","tripod-31-846","tripod-31-848","tripod-35-970","tripod-35-972","tripod-36-1003","tripod-36-1005"],"sample_id":["tripod-30-813","tripod-30-815","tripod-31-846","tripod-31-848","tripod-35-970","tripod-35-972","tripod-36-1003","tripod-36-1005"],"cell_subset":["amTreg","nTreg","amTreg","nTreg","amTreg","nTreg","nTreg","amTreg"],"sex":["F","F","M","M","F","F","M","M"],"nSequences":[386373,574469,850390,596005,933273,949332,729071,1134587],"V":[98,99,98,95,95,100,96,99],"J":[43,44,43,43,43,43,43,43],"VJ":[3540,3692,3629,3680,3345,3705,3026,3104],"ntCDR3":[64180,110601,84910,111238,42605,107506,26863,26721],"aaCDR3":[44097,69043,55848,69494,30781,67215,20428,20582],"aaClone":[57670,96796,75041,97371,38659,94066,24988,24752],"ntClone":[64180,110601,84910,111238,42605,107506,26863,26721],"chao1":[80147.43,127377.37,96934.79,129285.32,49657.34,115369.45,30108.53,29426.78],"iChao":[84110.14,131571.37,99886.65,133796.37,50706.05,117328.87,30674.24,29973.08]},"columns":[{"id":".rownames","name":"","type":"character","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"},"sortable":false,"filterable":false,"rowHeader":true},{"id":"sample_id","name":"sample_id","type":"factor","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":150,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"cell_subset","name":"cell_subset","type":"factor","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"sex","name":"sex","type":"factor","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"nSequences","name":"nSequences","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"V","name":"V","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"J","name":"J","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"VJ","name":"VJ","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"ntCDR3","name":"ntCDR3","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"aaCDR3","name":"aaCDR3","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"aaClone","name":"aaClone","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"ntClone","name":"ntClone","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"chao1","name":"chao1","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"iChao","name":"iChao","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}}],"bordered":true,"dataKey":"a0065a140752ed682fded4328bdd74d4"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script><ul>
<li>
<strong>otherData</strong>: a list encompassing the output of a
defined set of functions.</li>
</ul>
<p>This slot can be extracted using the <code><a href="../reference/RepSeqExperiment-class.html">oData()</a></code>
function.</p>
<p>For instance, if the <code>outFiltered</code> parameter was set to
TRUE during the RepSeqExperiment generation, filtered sequences would
have been stored in this slot and can be viewed as followed.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RepSeqExperiment-class.html">oData</a></span><span class="op">(</span><span class="va">RepSeqData</span><span class="op">)</span><span class="op">$</span><span class="va">filtered</span></span></code></pre></div>
<p>In addition, <code>oData</code> contains a list of colors attributed
for each group in the metadata file</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RepSeqExperiment-class.html">oData</a></span><span class="op">(</span><span class="va">RepSeqData</span><span class="op">)</span><span class="op">$</span><span class="va">label_colors</span></span>
<span><span class="co">#&gt; $sample_id</span></span>
<span><span class="co">#&gt;  tripod-30-813  tripod-30-815  tripod-31-846  tripod-31-848  tripod-35-970 </span></span>
<span><span class="co">#&gt;      "#7FC97F"      "#BEAED4"      "#FDC086"      "#FFFF99"      "#386CB0" </span></span>
<span><span class="co">#&gt;  tripod-35-972 tripod-36-1003 tripod-36-1005 </span></span>
<span><span class="co">#&gt;      "#F0027F"      "#BF5B17"      "#666666" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $cell_subset</span></span>
<span><span class="co">#&gt;    amTreg     nTreg </span></span>
<span><span class="co">#&gt; "#7FC97F" "#BEAED4" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sex</span></span>
<span><span class="co">#&gt;         F         M </span></span>
<span><span class="co">#&gt; "#FDC086" "#FFFF99"</span></span></code></pre></div>
<ul>
<li>
<strong>History</strong>: a data frame registering all operations
performed on the RepSeqExperiment object, such as the filtering and the
down-sampling.</li>
</ul>
<p>This slot can be extracted using the <code><a href="../reference/RepSeqExperiment-class.html">History()</a></code>
function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RepSeqExperiment-class.html">History</a></span><span class="op">(</span><span class="va">RepSeqData</span><span class="op">)</span></span></code></pre></div>
<div class="reactable html-widget html-fill-item" id="htmlwidget-febe03efa1a2d8d52a86" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-febe03efa1a2d8d52a86">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"history":["data directory=./extdata/MiAIRR","readAIRRSet; cores=2; fileFormat=MiAIRR; chain=TRA; ambiguous FALSE; unprod FALSE; filter.singletons FALSE; aa threshold=8"]},"columns":[{"id":"history","name":"history","type":"character","headerStyle":{"background":"#f7f7f8"}}],"bordered":true,"dataKey":"9d78e66e30b26e38ea1ee83505a5c191"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="repseqexp-merging">RepSeqExp merging<a class="anchor" aria-label="anchor" href="#repseqexp-merging"></a>
</h3>
<p><code><a href="../reference/mergeRepSeq.html">mergeRepSeq()</a></code> allows to combine two RepSeqExperiment
objects encompassing different datasets. As such, a same
<strong>sample_id</strong> cannot be in both datasets and should be
renamed before using this function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">'extdata/MiAIRR'</span><span class="op">)</span>,</span>
<span>                     package <span class="op">=</span> <span class="st">'AnalyzAIRR'</span><span class="op">)</span>,</span>
<span>                     full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dataset1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readAIRRSet.html">readAIRRSet</a></span><span class="op">(</span>fileList <span class="op">=</span> <span class="va">l</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                       cores <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>                       fileFormat <span class="op">=</span> <span class="st">"MiAIRR"</span>,</span>
<span>                       chain <span class="op">=</span> <span class="st">"TRA"</span>,</span>
<span>                       sampleinfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                       filter.singletons <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       outFiltered <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running on 1 cores...</span></span>
<span><span class="co">#&gt; Loading and filtering sequences...</span></span>
<span><span class="co">#&gt; Assembling sequences...</span></span>
<span><span class="co">#&gt; Creating a RepSeqExperiment object...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span></span>
<span><span class="va">dataset2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readAIRRSet.html">readAIRRSet</a></span><span class="op">(</span>fileList <span class="op">=</span> <span class="va">l</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                       cores <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>                       fileFormat <span class="op">=</span> <span class="st">"MiAIRR"</span>,</span>
<span>                       chain <span class="op">=</span> <span class="st">"TRA"</span>,</span>
<span>                       sampleinfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                       filter.singletons <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       outFiltered <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running on 1 cores...</span></span>
<span><span class="co">#&gt; Loading and filtering sequences...</span></span>
<span><span class="co">#&gt; Assembling sequences...</span></span>
<span><span class="co">#&gt; Creating a RepSeqExperiment object...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span></span>
<span><span class="va">dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeRepSeq.html">mergeRepSeq</a></span><span class="op">(</span>a <span class="op">=</span> <span class="va">dataset1</span>, b <span class="op">=</span> <span class="va">dataset2</span><span class="op">)</span></span></code></pre></div>
<div class="reactable html-widget html-fill-item" id="htmlwidget-1fb4450895fe099f74a1" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-1fb4450895fe099f74a1">{"x":{"tag":{"name":"Reactable","attribs":{"data":{".rownames":["tripod-30-813","tripod-30-815","tripod-31-846","tripod-31-848","tripod-35-970","tripod-35-972","tripod-36-1003","tripod-36-1005"],"sample_id":["tripod-30-813","tripod-30-815","tripod-31-846","tripod-31-848","tripod-35-970","tripod-35-972","tripod-36-1003","tripod-36-1005"],"nSequences":[386417,574542,850465,596086,933370,949481,729348,1134680],"V":[98,99,98,95,95,100,96,99],"J":[43,44,43,43,43,43,43,43],"VJ":[3540,3692,3629,3680,3345,3707,3026,3104],"ntCDR3":[64196,110624,84923,111265,42610,107535,26874,26725],"aaCDR3":[44112,69066,55861,69521,30786,67244,20439,20586],"aaClone":[57686,96819,75054,97398,38664,94095,24999,24756],"ntClone":[64196,110624,84923,111265,42610,107535,26874,26725],"chao1":[80170.8,127396.99,96946.52,129314.9,49665.03,115403.18,30121.45,29430.56],"iChao":[84135.19,131590.18,99898.12,133826.67,50713.99,117363.9,30687.55,29976.98]},"columns":[{"id":".rownames","name":"","type":"character","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"},"sortable":false,"filterable":false,"rowHeader":true},{"id":"sample_id","name":"sample_id","type":"factor","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":150,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"nSequences","name":"nSequences","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"V","name":"V","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"J","name":"J","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"VJ","name":"VJ","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"ntCDR3","name":"ntCDR3","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"aaCDR3","name":"aaCDR3","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"aaClone","name":"aaClone","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"ntClone","name":"ntClone","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"chao1","name":"chao1","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}},{"id":"iChao","name":"iChao","type":"numeric","format":{"cell":{"separators":true},"aggregated":{"separators":true}},"minWidth":200,"maxWidth":110,"align":"center","vAlign":"center","headerStyle":{"background":"#f7f7f8"}}],"bordered":true,"dataKey":"a0dda9d1ddb63495ffdf82d6425a5f0d"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vanessa Mhanna, Gabriel Pires, Grégoire Bohl, Karim El Soufi, Nicolas Tchitchek, David Klatzmann, Adrien Six, Hang-Phuong Pham, Encarnita Mariotti-Ferrandiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
